cryptarchia/cryptarchia.py

```diff
@@ -317,12 +321,29 @@ def __init__(self, genesis_state: LedgerState, config: Config):
         self.genesis_state = genesis_state
         self.ledger_state = {genesis_state.block.id(): genesis_state.copy()}
         self.epoch_state = {}
+        self.state = State.BOOTSTRAPPING
+        self.lib = genesis_state.block.id()  # Last immutable block, initially the genesis block
+
+    def to_online(self):
+        """
+        Call this method when the follower has finished bootstrapping. While this is somewhat left to implementations
+        https://www.notion.so/Cryptarchia-v1-Bootstrapping-Synchronization-1fd261aa09df81ac94b5fb6a4eff32a6 contains a great deal
+        of information and is the reference for the Rust implementation.
+        """
+        if self.state != State.BOOTSTRAPPING:
+            raise RuntimeError("Follower is not in BOOTSTRAPPING state")
+        self.state = State.ONLINE
+        self.update_lib()
 
     def validate_header(self, block: BlockHeader):
         # TODO: verify blocks are not in the 'future'
         if block.parent not in self.ledger_state:
             raise ParentNotFound
 
+        if not is_ancestor(self.lib, block.parent, self.ledger_state):
+            # If the block is not an ancestor of the last immutable block, we cannot process it.
+            raise ImmutableFork
+
```

Comment:
Isn't the `block.parent not in self.ledger_state`  check above sufficient?

We prune the block tree so that all blocks that we hold are descendents of LIB, if the parent is in our block tree, then the parent is a descendent of LIB, therefore this block is a descendent of LIB.

Reply:
oh right, good point.

So maybe the check should be
1. parent is in the block tree
2. parent height >= LIB height -- parent is forking at or after LIB